lib/elixir/lib/calendar/iso.ex

```diff
@@ -715,6 +715,89 @@ defmodule Calendar.ISO do
     {0, 1}
   end
 
+  @doc """
+  Implements the inspect protocol delegated from
+  `Date`, `Time`, `DateTime` and `NaiveDateTime`.
+
+  """
+  @doc since: "1.10.0"
+  @impl true
+  @spec inspect(
+          Calendar.date() | Calendar.time() | Calendar.datetime() | Calendar.naive_datetime(),
+          Inspect.Opts.t()
+        ) :: String.t()
+  def inspect(%Date{} = date, _) do
```

Comment:
Should the calendar be checked in all these clauses?

```suggestion
  def inspect(%Date{calendar: __MODULE__} = date, _) do
```

Reply:
Nevermind. Inspect accepts any struct/map with the proper fields.

Reply:
🤔 Or not...`:calendar` is part of those fields.

Btw, according to the spec, this should work:

```elixir
my_date = %{year: 2019, month: 06, day: 11, calendar: Calendar.ISO}
Calendar.ISO.inspect(my_date)
```

Reply:
This is an implementation of the function `inspect/2` supporting the `Inspect` protocol.  Therefore since its being called from the protocol as `calendar.inspect` I made the assumption that matching on the calendar wasn't necessary - since it would only be dispatched here if the calendar was `Calendar.ISO`.

Reply:
@kipcole9 Yes, I understand this function was designed to be used by the `Inspect` protocol. However, it's now part of the public interface of the `Calendar.ISO` module, and I think we need to consider that users might call this function directly. 

In summary, we need to answer these questions:

1. The value given as first argument might have a different calendar, should `inspect/2` work with that value? _I think it should not work_
1. The value given as first argument might be a map or a custom struct. According to the typespec `Calendar.date() | Calendar.time() | Calendar.datetime() | Calendar.naive_datetime()`, any map or struct with the proper fields should work: 

    https://github.com/elixir-lang/elixir/blob/0c686aa124bb0a31ba4010a78836c84e4b55066e/lib/elixir/lib/calendar.ex#L72-L112

    What should be changed, the typespec or the implementation? _I think we should change the implementation to accept any map/struct with the proper fields_

Reply:
Got it, I'll update and push another commit to this PR.

Reply:
I pushed a commit that uses optimistic matching in line with the other calendar functions.

Reply:
@kipcole9 please revert the commit. :) Those are structs on protocol dispatch. By definition they will have all of those fields. As it was before was totally fine.

What @fertapric was saying is that we need to do:

    def inspect(%Date{calendar: __MODULE__, ...})

inside Calendar.ISO.

Reply:
Reverted the previous commit, added a match on `calendar: __MODULE__` on each of the clauses.

What would you suggest as the correct types for the `@spec`?